src/test/java/codeu/controller/RegisterServletTest.java

```diff
@@ -1,35 +1,101 @@
+    mockResponse = Mockito.mock(HttpServletResponse.class);
+    mockRequestDispatcher = Mockito.mock(RequestDispatcher.class);
+    Mockito.when(mockRequest.getRequestDispatcher("/WEB-INF/view/register.jsp")).thenReturn(mockRequestDispatcher);
+  }
+
+  @Test
+  public void testDoGet() throws IOException, ServletException {
+    registerServlet.doGet(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testDoPost_BadUsername() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("bad !@#$% username");
+
+    registerServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verifyZeroInteractions(mockResponse);
+    Mockito.verify(mockRequest).setAttribute("error", "Please enter only letters, numbers, and spaces.");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testDoPost_NewUser() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("Testertest");
+    Mockito.when(mockRequest.getParameter("password")).thenReturn("password");
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = Mockito.mock(PrintWriter.class);
+    Mockito.when(mockResponse.getWriter()).thenReturn(printWriter);    
+
+    UserStore mockUserStore = Mockito.mock(UserStore.class);
+    Mockito.when(mockUserStore.isUserRegistered("Testertest")).thenReturn(false);
+    registerServlet.setUserStore(mockUserStore);
+
+    HttpSession mockSession = Mockito.mock(HttpSession.class);
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
+
+    registerServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verify(printWriter).println("<p>Username: Testertest</p>");
+    Mockito.verify(printWriter).println("<p>Password: password</p>");
+    
+    ArgumentCaptor<User> userArgumentCaptor = ArgumentCaptor.forClass(User.class);
+
+    Mockito.verify(mockUserStore).addUser(userArgumentCaptor.capture());
+    Assert.assertEquals(userArgumentCaptor.getValue().getName(), "Testertest");
+
+    Mockito.verify(mockResponse).sendRedirect("/login");
+  }
+
+  @Test
+  public void testDoPost_ExistingUser() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("Testertest");
+    Mockito.when(mockRequest.getParameter("password")).thenReturn("password");
+
+    UserStore mockUserStore = Mockito.mock(UserStore.class);
+    Mockito.when(mockUserStore.isUserRegistered("Testertest")).thenReturn(true);
+    registerServlet.setUserStore(mockUserStore);
+
+    HttpSession mockSession = Mockito.mock(HttpSession.class);
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
+
+    registerServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verify(mockUserStore, Mockito.never()).addUser(Mockito.any(User.class));
```

Comment:
This is a great verification, because it ensures that the datastore doesn't get polluted with duplicate entries--nice work! :+1: